### ISSUE TITLE
pass methods and attributes through containers

### ISSUE THREAD

**Issue by orbeckst** [datreant/datreant.core#25]
I find writing
```python
G = Group(pathname)
names = [x.name for x in G.members]
```

tedious. Furthermore, if `G` contains other groups, it sort of breaks (or I would have to manually and recursively flatten the group).

Instead I'd like some magic to happen:
```python
names = G.names
status = G.categories['status']
bool_finished = (G.categories['status'] == "complete")
bool_protein = G.tags["protein"]
```

where lists (or np arrays) are produced with the appropriate content from the members.

Essentially, override `__getattr__()` and friends to pass all unknown methods and attributes down to the lower level and apply them if possible. Do this recursively. Possibly return nested lists. 

Not sure how to deal with a group that includes itself and to break infinite recursion. Possibly forbid it in the first place (would also please [Zermelo](https://en.wikipedia.org/wiki/Russell%27s_paradox)).

Admittedly a bit dirty because there are the objects own methods that you cannot shadow... so feel free to close this issue with wontfix if you want to keep things tidy ;-)

Possibly related to #23.

**Comment by dotsdl** [comment]
I wrestled with this for a while, and the conclusion I came to was that `Group` should put all elements dealing with member parsing into `Members`. Although annoying, it has a few advantages:
* a `Group` is a first-class `Treant`. It has its own concept of tags and categories, and it can store data in the same way as a normal `Treant`. There are no special meanings to these components for a `Group` that differ from other `Treants`.
* the API of `Members` can be shared with `Bundle`, practically in its entirety, and slicing either yields another `Bundle`. In fact, the codebase used by these is exactly the same. Moving some member-parsing components into the top-level `Group` namespace, though very convenient, makes this more difficult.

Perhaps changes like those suggested could be made that maintain some of these advantages? Perhaps also some of this pain could be helped when we get a `query` method for `Members` and `Bundle`?

I think it worth mentioning that some of this namespace separation is lightly violated in [`MDSynthesis.Sim`](https://github.com/datreant/MDSynthesis), by way of the top-level `universe` property. However, since that's more domain-specific and focused I think I'm more comfortable with this kind of 'dirtiness'.

**Comment by dotsdl** [comment]
By the way, we definitely need a `flatten` method. See #26.

**Comment by orbeckst** [comment]
Ok, instead of working in the `Group` name space, one could add something to `members` for the same functionality.

```python
names = G.members.name
forcefields = G.members.categories['forcefield']
bool_protein = G.members.tags['protein']
```

 Of course, when the attribute/method access is propagated downwards one also needs to try the `.members` inside any of the group's members first, in case it is another `Group`.

**Comment by dotsdl** [comment]
Oh...I think I'm misunderstanding something. Let me lay this out in an example:

```python
from datreant import *

# make a group with three treants
g1 = Group('stooges')
g1.members.add(Treant('moe'), Treant('larry'), Treant('curly'))

# set a category for each treant in the group
for member in g1.members:
    member.categories['loser'] = 'True'

# set a category for the group
g1.categories['originals'] = 'True'

g2 = Group('later stooges')
g2.members.add(g1, Treant('shemp'))
```
At this point I have a `Group`, 'later stooges', that has a `Group` called 'stooges' and a `Treant` called 'shemp' as members. Now, what should happen if I did:

```python
g2.members.categories['loser']
```
?

**Comment by orbeckst** [comment]
Probably raise a `KeyError` because one member does not have a `categories['loser']`. (That's what would happen if you manually did something along the lines of (using #26 ):
```python
losers = [x.categories['loser'] for x in g2.members.flatten()]
```

Or slightly more ambiguous but convenient for how I would use it: return `None` for anything that's missing (which I consider appropriate use of `None`):

```python
g2.members.categories['loser'] == [True, True, True, None]
g2.members.name == ['moe', 'larry', 'curly', 'shemp']
```

Admittedly, I am not 100% sure what is really the best way is to deal with nested groups. You could argue that you should rather return `[[True, True, True], None]` and then the user can flatten it herself although flat might be the least confusing and most convenient. Personally I don't see much use for the nested structures because I am thinking about these collections more in terms of a normalized database (flat and tagged rows). Ultimately I am interested in ways to quickly pull parts of dataframes (which are organized in the same way as the group) using e.g. boolean indexing. E.g.
```
b_gromos = G.members.categories['forcefield'] == "GROMOS96"
b_of = G.members.categories['conformation'] == "OF"
...
# data for all simulations run with GROMOS96 and in the OF conformation
gromos_of = df[b_gromos & b_of]
```

Btw, fancy/boolean indexing of `Group.members` does not work, does it, i.e. I can't use this approach to slice Groups or Bundles?

**Comment by dotsdl** [comment]
Fancy indexing does not work for slicing members at the moment, no. I think it's an easy fix. I'll make an issue so it doesn't get forgotten here.

**Comment by dotsdl** [comment]
The ability to easily filter `Bundles` and `Group` members by tag has just landed in `develop`. For example:

```python
import datreant.core as dtr

b = dtr.Bundle(dtr.Treant('groot', tags=['amber', 'bark', 'sap']),
               dtr.Treant('sprout', tags=['green', 'bark', 'sapling']))

# returns a Bundle of Treants with this tag
b.tags['bark']
> <Bundle([<Treant: 'groot'>, <Treant: 'sprout'>])>

# returns a Bundle of Treants with ANY of these tags
# tuple functions as an OR
b.tags['green', 'bark']
> <Bundle([<Treant: 'sprout'>, <Treant: 'groot'>])>

# returns a Bundle of Treants that each have ALL of these tags
# list functions as an AND
b.tags[['green', 'bark']]
> <Bundle([])>

# lists and tuples can be nested for arbitrary OR and AND
b.tags[['green', ('bark', 'amber')], 'sap']
> <Bundle([<Treant: 'groot'>])>
```
This scheme doesn't have a way to say `not`, but perhaps we can come up with a clever way to make that work, too. Of course, this is meant to be a simple filtering scheme based on presence, not absence, of tags in particular combinations, so it's not essential we go further with it. We do eventually want a string-based selection system for `Treant` attributes, but that's a ways off.

**Comment by dotsdl** [comment]
Working on a similar `AggLimb` for `Categories`. That one will require a bit more thought.

**Comment by dotsdl** [comment]
See #43. Once that one is finished I think we'll have pretty much resolved this issue.

**Comment by orbeckst** [comment]
Closing, #43 is done and I think the UI has sufficiently stabilized that there's little point in trying to alter it in the way I outlined.
